### DISCUSSION TITLE
Dictionary Support

### DISCUSSION THREAD

**Discussion by EmilyBourne** [pyccel/pyccel#1053] [Ideas]
Dictionaries are very useful in python. Historically dictionaries were not supported due to [sympy restrictions](https://github.com/pyccel/pyccel/blob/master/pyccel/errors/messages.py#L90). As Pyccel is no longer constructed from sympy this is no longer a problem. Thus the only thing preventing implementation is the choice of how to implement it.

Personally I think the best solution would be to use an existing library however I have no experience with such libraries in Fortran or C.

From Google here are some options:

## Fortran (see [Fortran wiki](https://fortranwiki.org/fortran/show/Hash+tables))

- FLIBS
- fdict
- FPL
- fhash (object oriented)
- fhash (fpm package)
- ffhash

## C (see [stack overflow](https://stackoverflow.com/questions/4384359/quick-way-to-implement-dictionary-in-c))

- [uthash](https://troydhanson.github.io/uthash/)
- [GLIB](https://docs.gtk.org/glib/)
- [GNULIB](https://www.gnu.org/software/gnulib/)

**Comment by EmilyBourne** [comment]
Dictionaries will also require a clear type annotation.

Should we wish to support statements such as:
```
a = {}
a['test'] = 1
```
then the initial declaration may also need an annotation

**Reply by EmilyBourne** [reply]
Although maybe it is possible to leave the type determination until the first insertion as the elements cannot be accessed before then anyway

**Reply by yguclu** [reply]
Why do you think that a type annotation is needed?

**Reply by EmilyBourne** [reply]
So that they can be passed as arguments or stored in classes once classes

**Reply by yguclu** [reply]
You are thinking of a dictionary as syntactic sugar around a structure/class. I am not sure that is very useful for the Python user: if they need an immutable object they can use a `namedtuple`, which we could support quite easily; and if they need a mutable dictionary where the keys cannot be modified a standard `class` sounds more appropriate.

**Reply by EmilyBourne** [reply]
I think it is useful if only because it is one of the most basic Python objects. Most users learn about dictionaries before thy learn about `namedtuple`s, or even classes.
Is there another way to think about dictionaries?

**Reply by yguclu** [reply]
It is going to be very complicated to declare nested dictionaries, which is something that is commonly used.

In my opinion we should think of dictionaries as containers to which `(key, value)` pairs can be added dynamically; I am not convinced that it is useful to support them otherwise...

**Reply by EmilyBourne** [reply]
I agree. The only restriction that I would impose would be that all the keys should have the same type and all the values should have the same type.

Originally when I wrote this comment I was thinking about the code :
```
a = {}
```
which does not provide enough information for Pyccel to deduce the type of the dictionary (including the type of the keys and values). However as mentioned above it is maybe not a problem as the dictionary cannot be accessed until something is added anyway

**Reply by yguclu** [reply]
> I agree. The only restriction that I would impose would be that all the keys should have the same type and all the values should have the same type.

Sounds good to me. That appears to be the same assumption used by Numba: https://numba.pydata.org/numba-doc/dev/reference/pysupported.html#typed-dict

**Reply by EmilyBourne** [reply]
This should be doable with type annotations. See https://mypy.readthedocs.io/en/stable/cheat_sheet_py3.html#useful-built-in-types

**Comment by EmilyBourne** [comment]
Should we ever add C++ to Pyccel, a dictionary would be easily translated to a [map](https://en.cppreference.com/w/cpp/container/map)

**Comment by adam-urbanczyk** [comment]
Hey same comment as in the list thread, see also https://github.com/P-p-H-d/c-stl-comparison#synthesis for additional C implementations. It would be great if you also supported sets.

**Reply by EmilyBourne** [reply]
Thanks for this link.  I will have a look into this when I get back off holiday. If we can use existing libraries then the implementation in Pyccel should be faster

**Reply by EmilyBourne** [reply]
Of the packages listed only M\*LIB and COLLECTIONS-C support lists (dynamic arrays), maps and sets. Of these 2 M\*LIB has a permissive BSD licence while COLLECTIONS-C has a lGPL3 licence. Based on this I think it is worth looking into M\*LIB to see if it is suited to our needs

**Reply by EmilyBourne** [reply]
Some benchmarks can be found here:
https://github.com/P-p-H-d/mlib/wiki/performance

M\*LIB seems to compare favourably (although it may be worth noting that both these lists are compiled by the author of M*LIB)

**Reply by adam-urbanczyk** [reply]
Same holds for STC and they also provide benchmarks. I think both MLIB and STC are viable options.

**Reply by EmilyBourne** [reply]
According to the table STC implements sets but not maps. I have never used any of these libraries.
@adam-urbanczyk do you have experience with STC? Do you know if they have implemented sets since MLIB made this synthesis?

**Reply by adam-urbanczyk** [reply]
https://github.com/stclib/STC#containers 
I don't beyond looking at the readmes and btw the table does mention unordered set and map AFAICT.

**Reply by EmilyBourne** [reply]
It seems that maybe they have an unordered map but not an ordered map, so I guess it depends how close we want to stay to the Python implementation

**Reply by adam-urbanczyk** [reply]
They do have an ordered map too according to the readme. Anyhows ordered map (~std::map) is not a hash table and AFAIR python uses hash tables.

**Reply by adam-urbanczyk** [reply]
So Python since 3.6 uses hash tables that maintain key insertion order. This is something else than ordered maps using trees and ordering by key. Iteration with key insertion order seems to be quite uncommon. From a user perspective this particular property is not very critical.

**Reply by EmilyBourne** [reply]
Hello again,
I finally had time to look at the different libraries. Here is the summary:

|         | [M*LIB](https://github.com/P-p-H-d/mlib) | [STC](https://github.com/stclib/STC) | [CMC](https://github.com/LeoVen/C-Macro-Collections) | [CTL](https://github.com/rurban/ctl) | [CC](https://github.com/srdja/Collections-C) | [GLIB](https://docs.gtk.org/glib/index.html) | [KLIB](https://github.com/attractivechaos/klib) |
|---------|-------|-----|-----|-----|----|------|------|
| dicts   |  :heavy_check_mark:  | :heavy_check_mark:  | :heavy_check_mark: | :heavy_check_mark:  | :heavy_check_mark:  |  :heavy_check_mark:   |  :heavy_check_mark:  |
| sets    |  :heavy_check_mark: | :heavy_check_mark:  | :heavy_check_mark: |  :heavy_check_mark:|  :heavy_check_mark: |  :heavy_check_mark:   |  :heavy_check_mark: |
| lists   | :heavy_check_mark:  |  :heavy_check_mark:  | :heavy_check_mark: |  :heavy_check_mark:| :heavy_check_mark:  |  :heavy_check_mark:  |  :heavy_check_mark:|
| string | :heavy_check_mark:  |  :heavy_check_mark:  | :x: |  :heavy_check_mark: | :x: | :heavy_check_mark:   |  :heavy_check_mark:   |
| Licence |  BSD2  | MIT  |   MIT  | MIT | LGPL3  |  LGPL2  | MIT |
| Stars   | 705   |  865 |  421   |  140 (1k on original fork)  | 2.6k |  68   |  3.9k  |
| Docs    |  ++  | ++  |  Some  |  ++  | Some | ++  |   +  |
| Devs    |  Unknown | Norwegian Climate researchers  |  Student/Software Engineer | Forked by senior dev from [CTL](https://github.com/glouw/ctl) developed by minecraft dev |  Unknown |   Gnome | Unknown   |

Eliminating libs with copy-left licences and less support we still have:
- M*LIB
- STC
- CTL
- KLIB


# Performance
The performance results from https://github.com/P-p-H-d/mlib/wiki/performance (written by the M*LIB devs) show:

Type | Linked List | Array | Tree | Dict S | Dict B | Sort
-- | -- | -- | -- | -- | -- | --
M*LIB | **572 ms** | **312 ms** | **296 ms** | **32 ms** | **704 ms** | 888 ms
STC | 584 ms | 324 ms | 908 ms | 104 ms | 640 ms | 1304 ms
CTL | 592 ms | **312 ms** | 1204 ms | 460 ms | NA | 1256 ms
KLIB | 812 ms | **312 ms** | 436 ms | 56 ms | 756 ms | **784 ms**

*Tree means "Totally Sorted Tree": Red-Black Tree, AVL Tree, B-Tree or equivalent.
*Dict S: Dictionary with 8 bytes type,
*Dict B: Dictionary with 256 bytes type .

# Vector/List Code Examples
## M*LIB
```c
#include <stdio.h>
#include <gmp.h>
#include "m-array.h"

#define MPZ_OUT_STR(stream, x) mpz_out_str(stream, 0, x)
ARRAY_DEF(array_mpz, mpz_t,                                           \
          (INIT(mpz_init), INIT_SET(mpz_init_set), SET(mpz_set), CLEAR(mpz_clear), OUT_STR(MPZ_OUT_STR)))
	
int main(void) {
  array_mpz_t a;
  array_mpz_init(a);
  mpz_t x;
  mpz_init_set_ui(x, 16);
  array_mpz_push_back (a, x);
  mpz_set_ui(x, 45);             
  array_mpz_push_back (a, x);
  mpz_clear(x);
  printf ("ARRAY is: ");
  array_mpz_out_str(stdout, a);
  printf ("\n");
  printf ("First element is: ");
  mpz_out_str(stdout, 0, *array_mpz_get(a, 0));
  printf ("\n");
  array_mpz_clear(a);
}
```

## STC
```c
#define i_key int
#include "stc/cvec.h"

#include <stdio.h>

int main(void)
{
    // Create a vector containing integers
    cvec_int vec = {0};

    // Add two integers to vector
    cvec_int_push(&vec, 25);
    cvec_int_push(&vec, 13);

    // Append a set of numbers
    c_forlist (i, int, {7, 5, 16, 8})
        cvec_int_push(&vec, *i.ref);

    printf("initial:");
    c_foreach (k, cvec_int, vec) {
        printf(" %d", *k.ref);
    }

    // Sort the vector
    cvec_int_sort(&vec);

    printf("\nsorted:");
    c_foreach (k, cvec_int, vec) {
        printf(" %d", *k.ref);
    }
    cvec_int_drop(&vec);
}
```

## CTL
```c
#include <stdio.h>

#define POD
#define T int
#include <ctl/vector.h>

int compare(int* a, int* b) { return *b < *a; }

int main(void)
{
    vec_int a = vec_int_init();
    vec_int_push_back(&a, 9);
    vec_int_push_back(&a, 1);
    vec_int_sort(&a, compare);
    foreach(vec_int, &a, it)
        printf("%d\n", *it.ref);
    vec_int_free(&a);
}
```

## KLIB
```c
#include "kvec.h"

int main(void)
{
    kvec_t(int) array;
    kv_init(array);
    for (j = 0; j < N; ++j)
        kv_push(int, array, j);
    kv_destroy(array);
```

# Other comments
I am not convinced that `CTL` is the best option as the original fork has much less support, the new fork is much less heavily used and the fact that a fork was needed to keep the project active is not reassuring.

M*LIB has 3 dictionary types, depending on the complexity of the object being stored. This may add unnecessary complexity.

KLIB includes a math library with pseudo-random generators.

We will need to be careful to undefine macros to avoid problems later in the code.

KLIB support for lists seems more limited (no sort function).

STC allows contains an ND-array implementation (https://github.com/stclib/STC/blob/master/docs/cspan_api.md). It would be interesting to compare this to our implementation to see if we can find clues for https://github.com/pyccel/pyccel/discussions/1129. If it is faster and we choose this library then we could directly replaced our implementation with this one.

# My Preferences

Personally I find the M*LIB code harder to read than the others. It seems to be faster (according to their benchmarks) but not by much. I also think good docs are important and I'm wary of CTL having already been abandoned and reforked. My preference is therefore for STC. That said, I think any of these 4 would be valid solutions.

What do you think @yguclu @adam-urbanczyk ?

**Reply by adam-urbanczyk** [reply]
Thanks for coming back to this. I don't have hands on experience here, but going by activity and stars I'd choose either M*LIB or STC. Out of those two STC has a nicer API and better docs. So +1 for STC.

**Comment by EmilyBourne** [comment]
See https://fortran-lang.org/packages/data-types/ for existing Fortran implementations.

I like the look of https://github.com/SCM-NV/ftl as it seems to implement many objects (dict, list, etc) but would only add one dependency

**Reply by yguclu** [reply]
As we have encountered in the past with other third party libraries, I can see an important issue in the GPL license of this library which may be incompatible with the user code.

**Reply by EmilyBourne** [reply]
I think I see what you mean.

Users will not be able to distribute their translated code (which depends on the GPL licensed code) in a closed source format.

This would not be a problem for Pyccel itself as we would not have a dependency to the project. But the generated code would.

Is this what you mean?

Are you familiar enough with the different licences to know which kind we should aim for? The link to the summary page includes licences in the summary.

**Reply by EmilyBourne** [reply]
Actually no. This is LGPL3 not GPL3 so there is no copy-left if the library is only linked (https://en.m.wikipedia.org/wiki/GNU_Lesser_General_Public_License). I think this is how we would use. I don't expect us to modify their code at all (unless this was done by contributing to their project) so I don't see where the conflict could arise.

> The license allows developers and companies to use and integrate a software component released under the LGPL into their own (even proprietary) software without being required by the terms of a strong copyleft license to release the source code of their own components. However, any developer who modifies an LGPL-covered component is required to make their modified version available under the same LGPL license. For proprietary software, code under the LGPL is usually used in the form of a shared library, so that there is a clear separation between the proprietary and LGPL components.

**Reply by adam-urbanczyk** [reply]
It would be nice to be able to statically link without copy-left also for the fortran backend.

**Reply by EmilyBourne** [reply]
| | coretran | fdict | FTL | qContainers | gFTL |
|-|---|---|---|---|---|
| Lists | :heavy_check_mark: | :x: | :heavy_check_mark: |  :heavy_check_mark: |  :heavy_check_mark: |
| Dicts | :x: | :heavy_check_mark: | :heavy_check_mark: | :heavy_check_mark: |  :heavy_check_mark: |
| Sets | :x: | :arrow_up: | :heavy_check_mark: | :heavy_check_mark: |  :heavy_check_mark: |
| Strings | :heavy_check_mark: | :x: | :heavy_check_mark: | :x: | :x: |
| Licence | BSD 3 | MPL2 | lGPL 3 | BSD | Apache 2 |
| Dynamically linked copy left | :x: | :x: | :x: | :x: | :x: |
| Statically linked copy left | :x: | :heavy_check_mark: | :heavy_check_mark: | :x: | :x: |
| Last commit | Feb 2021 | Jun 2023 | Jun 2023 | Sep 2018 | Apr 2023 |
| Devs | PhD student (US) | computational physisict at TU Denmark| Software company specialising in Chemistry (SCM) | Unknown dev | computational scientist at NASA |

## Considerations

1.  **Quality** - a dedicated code, especially one written by experts is much more likely to be performant/bug free than one written into Pyccel (by beginners. I have never studied hashes and the interns have never used Fortran)
2.  **Maintenance** - a dedicated code used outside Pyccel is more likely to have related bug reports and prioritise fixing these
3.  **Licence Permissivity** - Pyccel's licence is MIT. For our users the more permissive the better
4.  **Ease of installation** - all options require installation from source in addition to Pyccel 
5.  **Keep number of dependencies low** - ftl and qContainers handle multiple use cases. A homegrown implementation would not add dependencies

**Reply by EmilyBourne** [reply]
I have tried to summarise the options shown on fortran-lang.org. I know that the maintainers of this website are hoping that having an index of existing packages will help encourage the use of Fortran. So I think the list ought to be fairly complete (additionally there is https://github.com/fortran-lang/stdlib which I haven't mentioned as it is hard to work out what is supported currently (there are no user docs) but it has an MIT license so maybe this should be something to bear in mind for a future iteration).

For me points 1 and 2 are the most important which is why I still like the look of FTL best. As regards licenses  in order to avoid copyleft and use a maintained project, it seems we would have to implement our own solution.

I think we should therefore do one (or both) of the following:
1. Use FTL as a temporary solution to allow users who don't want to distribute their code to translate more things
2. Contribute to the Fortran standard library until it has sufficient support for what we need

**Reply by adam-urbanczyk** [reply]
Nice overview! `qContainers` seems to wrap a c library, which _might_ be an issue for users. Also this looks like a possible candidate https://github.com/Goddard-Fortran-Ecosystem/gFTL with more permissive licensing.

**Reply by EmilyBourne** [reply]
> `qContainers` seems to wrap a c library, which _might_ be an issue for users.

I would be more inclined to see this as a benefit. The C library is maintained which makes it much less worrying that `qContainers` itself hasn't been touched on the last 5 years

> Also this looks like a possible candidate https://github.com/Goddard-Fortran-Ecosystem/gFTL with more permissive licensing.

Actually this looks great. (As explained in their README) It seems very similar to FTL but with a more permissive license. I think this could be a good solution that would cover most of the cases we need for now

**Comment by yguclu** [comment]
> Actually no. This is LGPL3 not GPL3 so there is no copy-left if the library is only linked (https://en.m.wikipedia.org/wiki/GNU_Lesser_General_Public_License). I think this is how we would use. I don't expect us to modify their code at all (unless this was done by contributing to their project) so I don't see where the conflict could arise.
> 
> > The license allows developers and companies to use and integrate a software component released under the LGPL into their own (even proprietary) software without being required by the terms of a strong copyleft license to release the source code of their own components. However, any developer who modifies an LGPL-covered component is required to make their modified version available under the same LGPL license. For proprietary software, code under the LGPL is usually used in the form of a shared library, so that there is a clear separation between the proprietary and LGPL components.

You are correct. FTL is released under LGPL license, not GPL as I wrote above. Therefore we can provide the user with a C Python extension which is dynamically linked to an FTL shared library.

**Reply by yguclu** [reply]
I suppose we can compile an FTL shared library when we install Pyccel. This should only require a Fortran 2003 compiler.

**Reply by EmilyBourne** [reply]
We would have to double check the licences. From what I have seen there are often specific clauses about distributing the software
